gh-107017: Rework the Fibonacci example

TommyUnreal opened the pull request:
Rework the example to be simple to follow and also the concepts needed for that are now explained one by one with their own shorter examples. 

Needs proofread. Please comment with your ideas for improvement.

<!-- gh-issue-number: gh-107017 -->
* Issue: gh-107017
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--107132.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

TommyUnreal committed:
Draft. Rework ´the Fibonacci example.

Rework the example to be simple to follow and also the concepts needed
for that are now explained one by one with their own shorter examples.

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,74 +478,103 @@ example::
      >>> a, b = 0, 1
      >>> while a < 1000:
-     ...     print(a, end=',')
+     ...     print(a, end=', ')
```
Do we want the space in `end` in this example? If so, update the output on line 546 with spaces after commas, so doctest passes on the CI.

TommyUnreal replied:
Thank you for your feedback.  I did this to showcase a different output style compared to the previous example, which used one character. My hope it might support a desire to experiment further with the end keyword. Maybe we might come up with a better way to do it. Or just not mess with it here 😂.

TommyUnreal replied:
In 1fae12b I tried a bit more obvious "two characters" and concepts from another example to create maybe a more interesting final example. Let me know what you think.

TommyUnreal committed:
Proofread, fixed the PEP 8, doctest.

TommyUnreal committed:
Fix rst formatting and missing literal block

TommyUnreal committed:
Fix inline literal.

TommyUnreal committed:
Better multichar end arg example in print().

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+
+A *multiple assignment* of variables us allows to set values of more than one
+variable on a single line.  Notice that if the value assignment is a
+expression, it is first evaluated and then assigned.::
```
`.::` will produce a full-stop _and_ a colon:

> .:

One or the other. I suggest losing the full-stop.

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+------------------
+
+The :keyword:`while` loop is example of a repeating cycle.  It performs
+a block of code over as long as the condition (in example: ``a < 10``) is
```
"in example" > "for example"

TommyUnreal replied:
What I wanted to say is that "in the example below the condition is the following expression `a < 10`", but that is no longer present in the examples. I will rework that.

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+   >>> count  # when count reached value of 5, while loop finished
+   5
+
+Note that block inside the while loop, or *body* of the loop is *indented*.
```
I suggest using ``while``, to draw attention to it.

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+the Python shell, you need to type a tab or space(s) for each indented line.
+Each line within a block must be indented by the same amount.
+When we want to exit the indented block of code in the Python shell, we must
+follow it by a blank line to indicate its completion.  This way, parser knows
```
"... the parser"

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+Each line within a block must be indented by the same amount.
+When we want to exit the indented block of code in the Python shell, we must
+follow it by a blank line to indicate its completion.  This way, parser knows
+we have finished typing the last line).::
```
Again, here we have a full-stop _and_ a colon in the output.

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+   ...     print(count, end=' ')  # end the output of print() with a whitespace
+   ...     count = count - 1  # lower the value by one with each repetition
+   ...
+   10 9 8 7 6
```
This is an elegant touch.

TommyUnreal replied:
Thank you :)

evildmp commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,118 @@ example::
+   ...      print('', a, '', end='->')
+   ...      a, b = b, a+b
+   ...
+    0 -> 1 -> 1 -> 2 -> 3 -> 5 -> 8 -> 13 -> 21 -> 34 -> 55 -> 89 -> 144 ->
```
And I like the way that the example illustrates the _value_ of using `end` - no need to explain, the example does it much more effectively. This will stick in someone's mind.

TommyUnreal committed:
Fix review hints. Explain condition better.

Fix parts where a paragraph ended both with full-stop and colon.
Added new paragraph, where example of simple loop is explained in
detail. Plus warning about indefinite loops.

TommyUnreal committed:
Fix inline literals.

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+
+   >>> a, b = 0, 1
+   >>> while a < 150:
+   ...      print('', a, '', end='->')
```
I understand that this utilities the print sep param, however this is nor explained nor inferred, so to a beginner it may be confusing 

Perhaps it could be beneficial to show off f-strings like 
```suggestion
   ...      print(f'{a} ->', end='')
```
Or to keep end as `->`
```suggestion
   ...      print(f'{a} ', end="->")
```

evildmp replied:
I very strongly prefer the original `print('', a, '', end='->')`, which reduces the punctuation the learner has to deal with to a minimum, and emphasises instead the connection between more easily assimilated things, like the English word "end" and the "arrow" ("->"). 

Showing off f-strings while demonstrating a different functionality introduces unnecessary cognitive burden. Two different kinds of quotes and two different kinds of brackets requires a lot of parsing and stopping. I don't think this is the right solution here.

TommyUnreal replied:
My reasoning: As this is an Introduction chapter, I wanted to keep the number of introduced concepts to the minimum. There are ways to code this better, but it is not formally wrong. The default separator is already shown in the example above and this just builds on that, so it's hopefully not confusing to the reader.

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+   >>> a, b = 0, 1
+   >>> while a < 150:
+   ...      print('', a, '', end='->')
+   ...      a, b = b, a+b
```
```suggestion
   ...      a, b = b, a + b
```
Looks better and I believe its what black would format

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+   >>> count = 10
+   >>> while count > 5:
+   ...     print(count, end=' ')  # end the output of print() with a whitespace
+   ...     count = count - 1  # lower the value by one with each repetition
```
```suggestion
   ...     count = count - 1  # decrement the value each iteration
```

evildmp replied:
I would strongly prefer "decrease the value by one".

The reason is that many people will not be familiar with the word "decrement", it's used very little in English outside technical contexts.

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+
+   >>> a, b = 1, 5  # multiple assignment of two variables
+   >>> a
    0
```
Needs to be `1`
![image](https://github.com/python/cpython/assets/23421705/900d91f6-44cd-46e8-9e27-b4567d91e9d8)

TommyUnreal replied:
Is this a doctest bug?

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+Repeating our code
+------------------
+
+The :keyword:`while` loop is example of a repeating cycle.  It performs
```
```suggestion
The :keyword:`while` loop is example of a repeating cycle. It performs
```

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+------------------
+
+The :keyword:`while` loop is example of a repeating cycle.  It performs
+a block of code over as long as the condition is
```
```suggestion
a block of code continuously as long as the condition is
```

evildmp replied:
I think that "over and over again" would be better than both these. "Continuously" is technically correct, but implies continuity, as in a continuous stream of water - whereas what we need to refer to is repeated execution of distinct, discrete blocks of code.

TommyUnreal replied:
Yes, this is more precise and clearer!

TommyUnreal committed:
Update Doc/tutorial/introduction.rst

Co-authored-by: Caeden Perelli-Harris <caedenperelliharris@gmail.com>

TommyUnreal committed:
Update Doc/tutorial/introduction.rst

Co-authored-by: Caeden Perelli-Harris <caedenperelliharris@gmail.com>

TommyUnreal committed:
Update Doc/tutorial/introduction.rst

More precise and clearer wording.

Co-authored-by: Caeden Perelli-Harris <caedenperelliharris@gmail.com>

TommyUnreal committed:
Fix multiple assignment example.

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+The :keyword:`while` loop is example of a repeating cycle. It performs
+a block of code continuously as long as the condition is
+fulfilled.  The test used in the example is a simple comparison you might
+know from arithmetics.  Fulfilled condition have the value ``True``.
```
```suggestion
know from arithmetics.  Fulfilled conditions have the value ``True``.
```

CaedenPH commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+a block of code continuously as long as the condition is
+fulfilled.  The test used in the example is a simple comparison you might
+know from arithmetics.  Fulfilled condition have the value ``True``.
+Unfulfilled condition have the value ``False``::
```
```suggestion
Unfulfilled conditions have the value ``False``::
```

CaedenPH reviewed (approved):
Other than these grammatical mistakes, looks like a good change to me

TommyUnreal committed:
Corrected grammar errors.

Co-authored-by: Caeden Perelli-Harris <caedenperelliharris@gmail.com>

TommyUnreal committed:
Better wording in example comment.

lower the value -> decrease the value

SylvainDe commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
-     0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,
+   4
+
+Function argumets
```
Typo here

TommyUnreal replied:
```suggestion
Function arguments
```

SylvainDe commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+Function argumets
+-----------------
+
+We already know :func:`print` function, that writes the value of the
```
We already know *the* print function ?

TommyUnreal replied:
```suggestion
We already know the : func:`print` function, which writes the value of the
```

SylvainDe reviewed (commented):
Very nice touch! Here are 2 cosmetic details

TommyUnreal committed:
Fixed typo, improved wording.

TommyUnreal commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+Function arguments
+-----------------
+
+We already know the : func:`print` function, which writes the value of the
```
```suggestion
We already know the :func:`print` function, which writes the value of the
```

TommyUnreal committed:
Fixed missing-space-before-role lint error.

TommyUnreal commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+   4
+
+Function arguments
+-----------------
```
```suggestion
------------------
```

TommyUnreal committed:
fIXED WARNING: Title underline too short lint.

hugovk committed:
Merge branch 'main' into 107017_tutorial_introduction_assumtions_removal_fibonacci

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+We can use Python for more complex tasks than adding two and two together.
+For instance, we can compute the numbers of the
+`Fibonacci series <https://en.wikipedia.org/wiki/Fibonacci_number>`_.
+To do that we need to utilize three following concepts.
```
```suggestion
To do that we need to use the three following concepts.
```

Let's use the [plain English](https://www.plainlanguage.gov/guidelines/words/use-simple-words-phrases/) "use" instead of "utilize".

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+Multiple Assignment
+-------------------
+
+A *multiple assignment* of variables us allows to set values of more than one
```
```suggestion
A *multiple assignment* of variables us allows to set the values of more than one
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+-------------------
+
+A *multiple assignment* of variables us allows to set values of more than one
+variable on a single line.  Notice that if the value assignment is a
```
```suggestion
variable on a single line.  Notice that if the value assignment is an
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+Repeating our code
+------------------
+
+The :keyword:`while` loop is example of a repeating cycle. It performs
```
```suggestion
The :keyword:`while` loop is an example of a repeating cycle. It performs
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+in this case, ``count < 5``.  To ensure that the loop finishes, we must make
+sure that the condition is *not fulfilled* at a certain step.  Otherwise, the
+code would repeat indefinitely.  To achieve that, we can increase the value
+of count.  As soon as the variable ``count`` reaches the value 5, the
```
```suggestion
of ``count``.  As soon as the variable ``count`` reaches the value 5, the
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+of count.  As soon as the variable ``count`` reaches the value 5, the
+condition will be ``False``::
+
+   >>> count = 0;  # define variable to which we will be adding 1 in a loop
```
```suggestion
   >>> count = 0  # define variable to which we will be adding 1 in a loop
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+condition will be ``False``::
+
+   >>> count = 0;  # define variable to which we will be adding 1 in a loop
+   >>> while count < 5: count = count + 1  # hit enter one more time to start the loop
+   ...
+   >>> count  # when count reached value of 5, while loop finished
+   5
+
+Note that block inside the ``while`` loop, or *body* of the loop is *indented*.
```
The example above uses a one-liner: `while count < 5: count = count + 1`

Let's use indentation in the example, to match the text. Something like this?


```suggestion
   >>> while count < 5:
   ...    count = count + 1  # hit enter one more time to start the loop
   ...
   >>> count  # when count reached value of 5, while loop finished
   5

Note that block inside the ``while`` loop, or *body* of the loop is *indented*.
```

hugovk commented on Doc/tutorial/introduction.rst:
```diff
@@ -478,73 +478,123 @@ example::
+
+We already know the :func:`print` function, which writes the value of the
+argument(s) it receives on screen.  The arguments are enclosed within
+parentheses ``()``. In simplest form, like ``print(a, b)``, the arguments
+are positional,  meaning the function processes them in the same order
```
```suggestion
parentheses ``()``. In its simplest form, like ``print(a, b)``, the arguments
are positional, meaning the function processes them in the same order
```

hugovk reviewed (commented):
A suggestion and some little wording things, otherwise looks good!
